In [7]:
import pandas as pd
import pandas as np
df = pd.read_csv('train_and_test2.csv')

In [8]:
print("Shape:", df.shape)
print("\nData types:\n", df.dtypes)
print("\nNulls per column:\n", df.isnull().sum())
print("\nDuplicate rows:", df.duplicated().sum())

Shape: (1309, 28)

Data types:
 Passengerid      int64
Age            float64
Fare           float64
Sex              int64
sibsp            int64
zero             int64
zero.1           int64
zero.2           int64
zero.3           int64
zero.4           int64
zero.5           int64
zero.6           int64
Parch            int64
zero.7           int64
zero.8           int64
zero.9           int64
zero.10          int64
zero.11          int64
zero.12          int64
zero.13          int64
zero.14          int64
Pclass           int64
zero.15          int64
zero.16          int64
Embarked       float64
zero.17          int64
zero.18          int64
2urvived         int64
dtype: object

Nulls per column:
 Passengerid    0
Age            0
Fare           0
Sex            0
sibsp          0
zero           0
zero.1         0
zero.2         0
zero.3         0
zero.4         0
zero.5         0
zero.6         0
Parch          0
zero.7         0
zero.8         0
zero.9         0
zero.10        0
z

Age has a missing values
Fare has 0 value
Sex  has 0 value

In [12]:
# fill Age with median
df['Age'] = df['Age'].fillna(df['Age'].median())

# fill Embarked with mode
df['Embarked'] = df['Embarked'].fillna(df['Embarked'].mode()[0])

# drop Cabin if mostly empty
df = df.drop(columns=['Cabin'])

KeyError: "['Cabin'] not found in axis"

In [13]:
before = df.shape[0]
df.drop_duplicates(inplace=True)
after = df.shape[0]
print(f"Removed {before - after} duplicate rows")

Removed 0 duplicate rows


In [16]:
print(df['Sex'].unique())

[0 1]


Sex is Pre-encode (0/1) rather than text labels.

In [18]:
# Cap outliers instead of removing them
df['Fare'] = df['Fare'].clip(lower=lower, upper=upper)

# Confirm it worked
Q1 = df['Fare'].quantile(0.25)
Q3 = df['Fare'].quantile(0.75)
IQR = Q3 - Q1
lower = Q1 - 1.5 * IQR
upper = Q3 + 1.5 * IQR
outliers_after = df[(df['Fare'] < lower) | (df['Fare'] > upper)]
print("Outliers after capping:", outliers_after.shape[0])

Outliers after capping: 0


In [17]:
Q1 = df['Fare'].quantile(0.25)
Q3 = df['Fare'].quantile(0.75)
IQR = Q3 - Q1
lower = Q1 - 1.5 * IQR
upper = Q3 + 1.5 * IQR
outliers = df[(df['Fare'] < lower) | (df['Fare'] > upper)]
print("Outliers found:", outliers.shape[0])

Outliers found: 171


171 outliers found in Fare capped rather than removed, since high fares are real first class fares and not data errors.

In [19]:
# Cap outliers instead of removing them
df['Fare'] = df['Fare'].clip(lower=lower, upper=upper)

# Confirm it worked
Q1 = df['Fare'].quantile(0.25)
Q3 = df['Fare'].quantile(0.75)
IQR = Q3 - Q1
lower = Q1 - 1.5 * IQR
upper = Q3 + 1.5 * IQR
outliers_after = df[(df['Fare'] < lower) | (df['Fare'] > upper)]
print("Outliers after capping:", outliers_after.shape[0])

Outliers after capping: 0


In [21]:
print(df.columns.tolist())

['Passengerid', 'Age', 'Fare', 'Sex', 'sibsp', 'zero', 'zero.1', 'zero.2', 'zero.3', 'zero.4', 'zero.5', 'zero.6', 'Parch', 'zero.7', 'zero.8', 'zero.9', 'zero.10', 'zero.11', 'zero.12', 'zero.13', 'zero.14', 'Pclass', 'zero.15', 'zero.16', 'Embarked', 'zero.17', 'zero.18', '2urvived']


In [22]:
for col in df.columns:
    print(col)

Passengerid
Age
Fare
Sex
sibsp
zero
zero.1
zero.2
zero.3
zero.4
zero.5
zero.6
Parch
zero.7
zero.8
zero.9
zero.10
zero.11
zero.12
zero.13
zero.14
Pclass
zero.15
zero.16
Embarked
zero.17
zero.18
2urvived


In [23]:
for col in df.columns:
    if 'urviv' in col.lower():
        print(repr(col))

'2urvived'


In [24]:
df['2urvived'] = df['2urvived'].astype('category')
df['Pclass'] = df['Pclass'].astype('category')
df['Sex'] = df['Sex'].astype('category')

print(df.dtypes)

Passengerid       int64
Age             float64
Fare            float64
Sex            category
sibsp             int64
zero              int64
zero.1            int64
zero.2            int64
zero.3            int64
zero.4            int64
zero.5            int64
zero.6            int64
Parch             int64
zero.7            int64
zero.8            int64
zero.9            int64
zero.10           int64
zero.11           int64
zero.12           int64
zero.13           int64
zero.14           int64
Pclass         category
zero.15           int64
zero.16           int64
Embarked        float64
zero.17           int64
zero.18           int64
2urvived       category
dtype: object


In [29]:
print("Rows:", df.shape[0])
print("Nulls now:\n", df.isnull().sum())
print("Duplicates now:", df.duplicated().sum())

Rows: 1309
Nulls now:
 Passengerid    0
Age            0
Fare           0
Sex            0
sibsp          0
zero           0
zero.1         0
zero.2         0
zero.3         0
zero.4         0
zero.5         0
zero.6         0
Parch          0
zero.7         0
zero.8         0
zero.9         0
zero.10        0
zero.11        0
zero.12        0
zero.13        0
zero.14        0
Pclass         0
zero.15        0
zero.16        0
Embarked       0
zero.17        0
zero.18        0
2urvived       0
dtype: int64
Duplicates now: 0


In [30]:
df.to_csv('cleaned_titanic.csv', index=False)

Nulls filled, outliers capped, dtypes fixed, duplicates checked.